feat: `SvelteKit Vite Plugin` support (#327)
vite-pwa/vite-plugin-pwa: docs/frameworks/sveltekit.md

# SvelteKit

::: warning
From SvelteKit version `1.0.0-next.358+`, SvelteKit is just another Vite plugin, and latest versions will also require you to update your application to use Vite 3.
:::

::: info
For `Type declarations`, `Prompt for update` and `Periodic SW Updates` go to [Svelte](/frameworks/svelte) entry.
:::

::: tip
You should remove all references to [SvelteKit service worker module](https://kit.svelte.dev/docs#modules-$service-worker) to disable it on your application.
:::

## SvelteKit Vite Plugin

`vite-plugin-pwa` provides the new `ViteSvelteKitPWA` plugin that will allow you to use `vite-plugin-pwa` in your SvelteKit applications.

> To update your project to use the new `vite-plugin-pwa` for SvelteKit, you only need to change the Vite config file (you don't need oldest `pwa` and `pwa-configuration` modules):

Reviewer: It'd probably make sense to write this aimed primarily towards people setting up new projects rather than those upgrading from older versions
```suggestion
To update your project to use the new `vite-plugin-pwa` for SvelteKit, you only need to add the plugin in the Vite config file (if you're upgrading from a prior version, you no longer need the `pwa` and `pwa-configuration` modules):
```

```ts
// vite.config.js

> // import { VitePWA } from 'vite-plugin-pwa'          <== replace this import
> import { ViteSvelteKitPWA } from 'vite-plugin-pwa' // <== with this import

Reviewer: We can just show how it should be configured. The replace suggestion won't apply for new users
```suggestion
import { ViteSvelteKitPWA } from 'vite-plugin-pwa'
```

import { sveltekit } from '@sveltejs/kit/vite'

/** @type {import('vite').UserConfig} */
const config = {
  plugins: [
    sveltekit(),
    ViteSvelteKitPWA({/* options */})
  ],
}

export default config
```

> ## Plugin Configuration
> 
> `vite-plugin-pwa` has been modified to automatically detect SvelteKit plugin: once detected, it will add a set of default configuration options to your `vite-plugin-pwa` options:
> - allows you to configure the SvelteKit build output folder: defaults to `.svelte-kit`, but you can change it with `svelteKitOptions.outDir` plugin option (([SvelteKit outDir](https://kit.svelte.dev/docs/configuration#outdir))).
> - configures the `globDirectory` with SvelteKit build output folder: `globDirectory: '.svelte-kit/output'` (using previous `svelteKitOptions.outDir` option).
> - adds `.svelte-kit/output/client` and `.svelte-kit/output/prerendered` folders to the `globPatterns`: `globPatterns: ['prerendered/**/*.html', 'client/**/*.{js,css,ico,png,svg,webp}']`.
> - configures default Rollup assets naming convention: `dontCacheBustURLsMatching: /-[a-f0-9]{8}\./` (by default, `vite-plugin-pwa` will use Vite assets naming convention: `/\.[a-f0-9]{8}\./`).
> - excludes adding manifest icons: `includeManifestIcons: false` (Vite will copy all `publicDir` content to the SvelteKit output folder before `vite-plugin-pwa` runs, and so, you will end up with duplicated entries in the service worker's precache manifest).
> - allows you to configure `trailingSlash` option: `vite-plugin-pwa` will use it in its internal Workbox `manifestTransform` callback ([SvelteKit trailingslash](https://kit.svelte.dev/docs/configuration#trailingslash)).
> - allows you to configure `fallback` adapter option: `vite-plugin-pwa` will configure it in the `workbox.navigateFallback` option, only when using `generateSW` strategy ([adapter-static fallback](https://github.com/sveltejs/kit/tree/master/packages/adapter-static#fallback)).
> 
> Some of the above options will be excluded if you already provide them or if you provide options where there may be a conflict between them: you can view the source code of the [SvelteKit PWA configuration module](https://github.com/antfu/vite-plugin-pwa/tree/main/src/integrations/sveltekit/config.ts) to verify that there are no conflicts.
> 

Reviewer: This is internal implementation details that users should not need to worry about. It makes it sound a lot more complicated than it really is to setup the plugin
```suggestion
```

## SvelteKit Pages

> If you want your application can work offline, you should remove `hydrate: false` from all your pages, it will prevent to inject the layout and so will not work offline.

Reviewer: content looks good. just a minor wording suggestion
```suggestion
If you want your application to work offline, you should ensure you have not set `hydrate: false` on any of your pages since it will prevent injecting JavaScript into the layout for offline support.
```

### Auto Update

Since SvelteKit uses SSR/SSG, we need to call the `vite-plugin-pwa` virtual module using a dynamic `import`.

The best place to include the virtual call will be in main layout of the application (you should register it in any layout):

::: details src/routes/__layout.svelte
```html
<script>
  import { onMount } from 'svelte'
  import { browser, dev } from '$app/env'
  import { assets } from '$app/paths'

  onMount(async () => {

>     if (!dev && browser) {

Reviewer: you probably don't need `&& browser`. `onMount` only runs on the browser and is generally tree-shaken away on the server, I believe
```suggestion
    if (!dev) {
```

## SvelteKit Vite Plugin

> `Sveltkit` maintainers have been working hard to align `Vite` and `SvelteKit`. The main changes done in `SvelteKit` plugin are:
> - exposes and configures `publicDir` properly  to allow `Vite` plugins using it: `publicDir` is configured with `config.kit.files.assets` (defaults to `static` folder).
> - exposes and configures `outDir` to allow `Vite` plugins using it: `outDir` is configured with `${svelteKitOutDir}/output/client` (defaults to `.svelte-kit/output/client` folder).
> - don't call `process.exit` in the `closeBundle` hook: which means that any `Vite` plugin that is configured after the `SvelteKit` plugin (or has `enforce: 'post'`) will be called by `Vite`.

Reviewer: It's very nice of you to mention all the hard work that's gone into this! I don't know if users will need it in a reference guide though. Perhaps we can highlight it on the SvelteKit monthly blog post instead
```suggestion
```

Author: done

> To update your project to use the new `vite-plugin-pwa` for `SvelteKit`, you only need to change the `Vite` config file (you don't need oldest `pwa` and `pwa-configuration` modules):

Reviewer: usually backticks are for code. it seems a bit unusual to me to list project names that way unless it's the repository name like `vite-plugin-pwa`
```suggestion
To update your project to use the new `vite-plugin-pwa` for SvelteKit, you only need to change the Vite config file (you don't need oldest `pwa` and `pwa-configuration` modules):
```

Some of the above options will be excluded if you already provide them or if you provide options where there may be a conflict between them: you can view the source code of the [SvelteKit PWA configuration module](https://github.com/antfu/vite-plugin-pwa/tree/main/src/integrations/sveltekit/config.ts) to verify that there are no conflicts.

## SvelteKit Pages

If you want your application can work in offline, you should remove `hydrate: false` from all your pages, this will prevent to inject the layout and so will not work in offline.

### Auto Update

> Since `SvelteKit` uses `SSR / SSG`, we need to call the `vite-plugin-pwa` virtual module using `dynamic import`.

Reviewer: ```suggestion
Since `SvelteKit` uses SSR / SSG, we need to call the `vite-plugin-pwa` virtual module using a dynamic `import`.
```

The best place to include the virtual call will be in main layout of the application (you should register it in any layout):

::: details src/routes/__layout.svelte
```html
<script>
  import { onMount } from 'svelte'
  import { browser, dev } from '$app/env'

  onMount(async () => {
    if (!dev && browser) {
      const { registerSW } = await import('virtual:pwa-register')
      registerSW({
        immediate: true,
        onRegistered(r) {
          // uncomment following code if you want check for updates
          // r && setInterval(() => {
          //    console.log('Checking for sw update')
          //    r.update()
          // }, 20000 /* 20s for testing purposes */)
          console.log(`SW Registered: ${r}`)
        },
        onRegisterError(error) {
          console.log('SW registration error', error)
        }
      })
    }
  })
</script>

<svelte:head>
  {#if (!dev && browser)}
    <link rel="manifest" href="/manifest.webmanifest">
  {/if}
</svelte:head>

<main>
  <slot />
</main>
```
:::

### Prompt for update

> Since `SvelteKit` uses `SSR / SSG`, we need to add the `ReloadPrompt` component using `dynamic import`. `vite-plugin-pwa` plugin will only register the service worker on build (check the [Development section](/guide/development)), it is aligned with the current behavior of [SvelteKit service worker module](https://kit.svelte.dev/docs#modules-$service-worker).

Reviewer: ```suggestion
Since `SvelteKit` uses SSR / SSG, we need to add the `ReloadPrompt` component using a dynamic `import`. `vite-plugin-pwa` plugin will only register the service worker on build (check the [Development section](/guide/development)), it is aligned with the current behavior of [SvelteKit service worker module](https://kit.svelte.dev/docs#modules-$service-worker).
```

```
:::

::: details $lib/ReloadPrompt.svelte
```html
<script lang="ts">
	import { useRegisterSW } from 'virtual:pwa-register/svelte'
	const {
		needRefresh,
		updateServiceWorker
	} = useRegisterSW({
		onRegistered(r) {

> 		    // uncomment following code if you want check for updates
> 			// r && setInterval(() => {
>             //    console.log('Checking for sw update')
>             //    r.update()
>             // }, 20000 /* 20s for testing purposes */)

Reviewer: ```suggestion
		// uncomment following code if you want check for updates
		// r && setInterval(() => {
		//    console.log('Checking for sw update')
		//    r.update()
		// }, 20000 /* 20s for testing purposes */)
```
